Implement user provided aggregation specs

ChristianZimpelmann opened the pull request:
### What problem do you want to solve?

- For user-written functions, it is currently impossible to specify aggregation variables in a similar way.

### Proposed implementation

- Add a new argument `aggregation_specs` to `compute_taxes_and_transfers`. 
- Elements of this dictionary are added to the aggregation specifications specified in the code base and override those if a key is used twice.

Closes #360 

### Todo

- [x] Implement new argument for `compute_taxes_and_transfers` as described in #360.
- [x] Write tests
- [x] Mention in tutorials

hmgaudecker commented:
Running advanced usage tutorial fails:

```pytb
[File /mnt/econ/gettsim/gettsim/gettsim/visualization.py:133, in plot_dag(functions, targets, columns_overriding_functions, check_minimal_specification, selectors, labels, tooltips, plot_kwargs, arrow_kwargs, edge_kwargs, label_kwargs, node_kwargs)
    ]()[131](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=130)[ # Create and add aggregation functions
    ]()[132](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=131)[ typical_data_cols = list(TYPES_INPUT_VARIABLES)
--> ]()[133](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=132)[ aggregation_funcs = _create_aggregation_functions(
    ]()[134](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=133)[     user_and_internal_functions, targets, typical_data_cols
    ]()[135](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=134)[ )
    ]()[136](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=135)[ all_functions = {**user_and_internal_functions, **aggregation_funcs}
    ]()[138](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=137)[ all_functions = {
    ]()[139](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=138)[     k: v for k, v in all_functions.items() if k not in columns_overriding_functions
    ]()[140](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=139)[ }

TypeError: _create_aggregation_functions() missing 1 required positional argument: 'user_provided_aggregation_specs']()
```

Aside: That was a misunderstanding/incomplete discussion yesterday re the basic usage. I'd prefer two households, maybe one single andone couple? The data should have the same form as in an actual application (multiple households).

hmgaudecker commented on gettsim/interface.py:
```diff
@@ -54,13 +55,16 @@ def compute_taxes_and_transfers(
+        the keys of the dictionary are used as a name instead of the function name. For
+        all other objects, the name is inferred from the function name.
+    aggregation_specs : dict, default None
+        A dictionary which contains specs for functions which aggregate variables on
```
Refer to docs?

hmgaudecker commented on gettsim/interface.py:
```diff
@@ -152,6 +168,9 @@ def check_data_check_functions_and_merge_functions(
     data : dict of pandas.Series
         Data provided by the user.
+    aggregation_specs : dict, default None
+        A dictionary which contains specs for functions which aggregate variables on
```
Again, refer to some place where I find useful information how this should look like. If I am lost before, I am also lost after reading this.

hmgaudecker commented on gettsim/interface.py:
```diff
@@ -815,7 +836,14 @@ def _create_aggregation_functions(user_and_internal_functions, targets, data_col
-    aggregation_dict = {**aggregation_dict, **automated_sum_aggregation_specs}
+
+    # Add automated aggregation specs.
+    # Note: For duplicate keys, automated specs are treated with lower priority.
```
```suggestion
    # Note: For duplicate keys, explicitly set specs are treated with higher priority than automated specs.
```
Better the positive version. More importantly, keep it the same way as next docstring.

hmgaudecker commented on gettsim/parameters/soz_vers_beitr.yaml:
```diff
@@ -783,10 +783,10 @@ ag_abgaben_geringf:
+    de: Mindestalter Beitragszuschlag für Kinderlose
+    en: Minimum age for additional contribution for childless
   description:
     de: Der Beitragssatz für Kinderlose steigt ab dieser Altersgrenze an.
```
```suggestion
    de: Der Beitragssatz für Kinderlose ist ab diesem Alter höher.
```

hmgaudecker commented on gettsim/parameters/soz_vers_beitr.yaml:
```diff
@@ -783,10 +783,10 @@ ag_abgaben_geringf:
+    en: Minimum age for additional contribution for childless
   description:
     de: Der Beitragssatz für Kinderlose steigt ab dieser Altersgrenze an.
     en: The contribution rate for childless people increases from this age limit.
```
```suggestion
    en: Childless individuals face a higher contribution rate starting at this age.
```

hmgaudecker commented on gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -254,7 +254,7 @@ def _vorsorgeaufw_vom_lohn_bis_2019_single(
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2019 for singles.
+    """Calculate vorsoge expenditures until 2019 for singles.
```
```suggestion
    """Calculate precautionary expenditures until 2019 for singles.
```
or "Vorsorge", but capitalised and without typo :smile:, same below.

hmgaudecker commented on gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -280,7 +280,7 @@ def _vorsorgeaufw_vom_lohn_bis_2019_tu(
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2019 per tax unit.
+    """Calculate vorsoge expenditures until 2019 per tax unit.
```
```suggestion
    """Calculate precautionary expenditures until 2019 per tax unit.
```

hmgaudecker commented on gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:
```diff
@@ -310,7 +310,7 @@ def _berechne_vorsorgeaufw_bis_2004(
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2004.
+    """Calculate vorsoge expenditures until 2004.
```
```suggestion
    """Calculate precautionary expenditures until 2004.
```

hmgaudecker reviewed (commented):
Excellent, thanks!!! 

Very minor stuff, except for the missing argument noted elsewhere.

ChristianZimpelmann commented:
Many thanks for the suggestions!

- I fixed the error in visualizations (and added a test for it). Advanced usage is running through now.
- I added a second household to basic usage.

hmgaudecker commented:
Somewhere the behaviour got lost that we show all missing columns, can we restore that?

![image](https://user-images.githubusercontent.com/158456/160792082-5ebf818a-c00d-40f4-aee3-04ef4d31129b.png)

(from advanced tutorial, else would need to adjust that)

hmgaudecker commented:
I fixed the issues in #365 in the previous commit. The notebooks `policy_functions` and `parameters` currently fail. I have no idea what somebody tried to do there, should never have been publicly visible. Please fix that, I won't manage today (maybe tomorrow).

Re `synthetic.py`: This is a good example of code that would work well to cook up a one-time project. It does not live up to standards for a long-term collaborative project (clear API, declarative instead of imperative code, ...). For the future: Please ensure quality standards before merging. Better to have a missing feature than an enduring nightmare. Sorry for not looking into it back then. Thanks.

ChristianZimpelmann commented:
- I adjusted the order of input data checks
- uploading code coverage reports is back in (got lost when commenting out jax tests)
- I fixed the problem in `synthetic.py` that made the tutorials fail (thx for fixing the issue)

btw.: Is there a better way to check the tutorials other than selecting each one and clicking "Run all Cells"?

hmgaudecker commented:
Thanks! 

> btw.: Is there a better way to check the tutorials other than selecting each one and clicking "Run all Cells"?

Should be possible to let them run as part of the CI or so, not sure. Only problem would be with the advanced_usage where we show errors, but maybe one could continue with execution despite them.

hmgaudecker reviewed (approved):
Wonderful, thanks!
